starboard/shared/starboard/player/filter/audio_discard_duration_tracker.cc

```diff
@@ -0,0 +1,106 @@
+// Copyright 2023 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "starboard/shared/starboard/player/filter/audio_discard_duration_tracker.h"
+
+#include <algorithm>
+
+#include "starboard/common/log.h"
+#include "starboard/shared/starboard/media/media_util.h"
+
+namespace starboard {
+namespace shared {
+namespace starboard {
+namespace player {
+namespace filter {
+using shared::starboard::media::AudioSampleInfo;
+
+void AudioDiscardDurationTracker::CacheDiscardDuration(
+    const scoped_refptr<InputBuffer>& input_buffer,
+    SbTime buffer_duration) {
+  SbTime discard_duration_from_front =
+      input_buffer->audio_sample_info().discarded_duration_from_front;
+  SbTime discard_duration_from_back =
+      input_buffer->audio_sample_info().discarded_duration_from_back;
+
+  ScopedLock lock(mutex_);
```

Comment:
This function should be private, and the lock should be held inside CacheMultipleDiscardDurations(), so we can ensure that the effect of calling CacheMultipleDiscardDurations() is consistent to the observers.